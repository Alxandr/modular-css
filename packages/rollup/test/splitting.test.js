/* eslint max-statements: "off" */
"use strict";

const { rollup } = require("rollup");

const { cwd, temp, find } = require("test-utils/fixtures.js");
const namer = require("test-utils/namer.js");

require("test-utils/expect-dir-snapshot.js");

const plugin = require("../rollup.js");

const error = (root) => {
    throw root.error("boom");
};

error.postcssPlugin = "error-plugin";

const experimentalCodeSplitting = true;
const assetFileNames = "assets/[name][extname]";
const chunkFileNames = "[name].js";
const format = "es";
const map = false;
const sourcemap = false;
const json = true;

describe("/rollup.js", () => {
    describe("code splitting", () => {
        it("should support splitting up CSS files", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,
                
                input : [
                    find("simple.js"),
                    find("dependencies.js"),
                ],

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                    }),
                ],
            });
    
            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,
                
                dir : temp(),
            });

            expect(temp("assets")).toMatchDirSnapshot();
        });

        it("should support splitting up CSS files w/ shared assets", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,

                input : [
                    find("css-chunks/a.js"),
                    find("css-chunks/b.js"),
                ],

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                    }),
                ],
            });

            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,

                dir : temp(),
            });

            expect(temp("assets")).toMatchDirSnapshot();
        });
        
        it("shouldn't put bundle-specific CSS in common.css", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,

                input : [
                    find("common-splitting/a.js"),
                    find("common-splitting/c.js"),
                ],

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                    }),
                ],
            });

            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,

                dir : temp(),
            });

            expect(temp("assets")).toMatchDirSnapshot();
        });

        it("should support manual chunks", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,

                input : [
                    find("manual-chunks/a.js"),
                    find("manual-chunks/b.js"),
                ],

                manualChunks : {
                    shared : [
                        find("manual-chunks/c.js"),
                    ],
                },

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                    }),
                ],
            });

            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,

                dir : temp(),
            });

            expect(temp("assets")).toMatchDirSnapshot();
        });

        it("should support dynamic imports", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,

                // treeshake : false,

                input : [
                    find("dynamic-imports/a.js"),
                    find("dynamic-imports/b.js"),
                ],

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                    }),
                ],
            });

            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,

                dir : temp(),
            });

            expect(temp("assets/")).toMatchDirSnapshot();
        });

        it("should ouput only 1 JSON file", async () => {
            const bundle = await rollup({
                experimentalCodeSplitting,

                input : [
                    find("simple.js"),
                    find("dependencies.js"),
                ],

                plugins : [
                    plugin({
                        cwd,
                        namer,
                        map,
                        json,
                    }),
                ],
            });

            await bundle.write({
                format,
                sourcemap,

                assetFileNames,
                chunkFileNames,

                dir : temp(),
            });

            expect(temp("assets")).toMatchDirSnapshot();
        });
    });
});
