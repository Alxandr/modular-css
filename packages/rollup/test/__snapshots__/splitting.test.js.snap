// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/rollup.js code splitting should ouput only 1 JSON file 1`] = `
Array [
  Object {
    "file": "common.css",
    "text": "/* simple.css */
.fooga {
    color: red;
}
",
  },
  Object {
    "file": "dependencies.css",
    "text": "/* dependencies.css */
.wooga {

    background: blue;
}
",
  },
  Object {
    "file": "exports.json",
    "text": "{
    \\"dependencies.css\\": {
        \\"wooga\\": \\"fooga wooga\\"
    },
    \\"simple.css\\": {
        \\"str\\": \\"\\\\\\"string\\\\\\"\\",
        \\"fooga\\": \\"fooga\\"
    }
}",
  },
]
`;

exports[`/rollup.js code splitting should support dynamic imports 1`] = `
Array [
  Object {
    "file": "a.css",
    "text": "/* dynamic-imports/a.css */
.a {

    color: aqua;
}
",
  },
  Object {
    "file": "b.css",
    "text": "/* dynamic-imports/b.css */
.b {
    color: blue;
}
",
  },
  Object {
    "file": "c.css",
    "text": "/* dynamic-imports/c.css */
.c {
    
    color: cyan;
}
",
  },
  Object {
    "file": "common.css",
    "text": "/* dynamic-imports/d.css */
.d {
    color: darkred;
}
",
  },
]
`;

exports[`/rollup.js code splitting should support manual chunks 1`] = `
Array [
  Object {
    "file": "a.css",
    "text": "/* manual-chunks/a.css */
.a {

    color: red;
}
",
  },
  Object {
    "file": "b.css",
    "text": "/* manual-chunks/b.css */
.b {

    color: blue;
}
",
  },
  Object {
    "file": "common.css",
    "text": "/* manual-chunks/d.css */
/* manual-chunks/c.css */
.c {
    color: green;
    background: darkred;
}
",
  },
]
`;

exports[`/rollup.js code splitting should support splitting up CSS files 1`] = `
Array [
  Object {
    "file": "common.css",
    "text": "/* simple.css */
.fooga {
    color: red;
}
",
  },
  Object {
    "file": "dependencies.css",
    "text": "/* dependencies.css */
.wooga {

    background: blue;
}
",
  },
]
`;

exports[`/rollup.js code splitting should support splitting up CSS files w/ shared assets 1`] = `
Array [
  Object {
    "file": "a.css",
    "text": "/* css-chunks/a.css */
.a {

    color: aqua;
}
",
  },
  Object {
    "file": "b.css",
    "text": "/* css-chunks/b.css */
.b {
    
    color: blue;
}
",
  },
  Object {
    "file": "chunk.css",
    "text": "/* css-chunks/c.css */
.c {

    color: cyan;
}
",
  },
  Object {
    "file": "common.css",
    "text": "/* css-chunks/shared.css */
.shared {
    color: snow;
}
",
  },
]
`;

exports[`/rollup.js code splitting shouldn't put bundle-specific CSS in common.css 1`] = `
Array [
  Object {
    "file": "a.css",
    "text": "/* common-splitting/shared.css */
.shared {
    color: snow;
}
/* common-splitting/a.css */
.a {

    color: aqua;
}
/* common-splitting/b.css */
.b {

    color: blue;
}
",
  },
  Object {
    "file": "c.css",
    "text": "/* common-splitting/c.css */
.c {
    color: cyan;
}
",
  },
]
`;
